[XLA:GPU] add cuDNN flash attention support in XLA (3rd PR with only rewriter changes)

Cjkkkk opened the pull request:
This is the 3nd PR of splitting https://github.com/openxla/xla/pull/5910 with only rewriter changes
1st PR https://github.com/openxla/xla/pull/6293 merged.
2nd PR https://github.com/openxla/xla/pull/6657 merged.

* Add pattern match for causal mask
* Add paxml dropout pattern match
* Add flash attention fusion
* Add flash attention support cuDNN version guard
* Add tests for flash attention rewriter/e2e

ddunl commented:
FYI 2nd PR is getting rolled back as we speak unfortunately, I'll have more details tomorrow (or maybe @cheshire can look into it if he has time tomorrow)

Cjkkkk commented:
> FYI 2nd PR is getting rolled back as we speak unfortunately, I'll have more details tomorrow (or maybe @cheshire can look into it if he has time tomorrow)



> FYI 2nd PR is getting rolled back as we speak unfortunately, I'll have more details tomorrow (or maybe @cheshire can look into it if he has time tomorrow)

Thanks for the info. Let me know once you have more details.

Cjkkkk commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -402,6 +410,93 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
+      auto name = comp->name();
+      auto entry_computation = mod->entry_computation();
+      bool is_causal_mask = true;
+      for (HloInstruction* instr :
```
Could you elaborate on this one? Could you point me to some implementation from fusion computations, collective call computations or custom call computations?

akuegel replied:
Here is the PR where the back pointers for collective call computations were added:

https://github.com/openxla/xla/pull/6171

gbaned replied:
Hi @Cjkkkk Any update on above [comment ](https://github.com/openxla/xla/pull/6872#discussion_r1392301103)form @akuegel . Thank you!

Cjkkkk replied:
Should we open a separate PR for this or just include this in the current PR?

akuegel replied:
A separate PR would probably be best.

Cjkkkk replied:
I created a new PR to address the issue discussed above: https://github.com/openxla/xla/pull/7179.

Cjkkkk commented:
@akuegel Hi Adrian, do we have any other items that need to be done for this PR since the back pointer stuff is merged?

akuegel commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -402,6 +410,81 @@ StatusOr<bool> IsSupportedBMM2(const HloInstruction* bmm_2,
+      std::vector<int64_t> seq_q_dims,
+      GetNonContractingDims(
+          bmm_1->operand(0)->shape(),
+          bmm_1->dot_dimension_numbers().lhs_batch_dimensions(),
```
I think it would be useful to use const reference variable to bmm_1->dot_dimension_numbers(), as this is used quite often in the following code.

akuegel commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -1205,38 +1326,66 @@ StatusOr<HloInstruction*> FuseFwdMultiHeadedAttentionBlock(
-    // TODO Flush attention will have a different shape in training.
     activation_output = bmm_2->mutable_operand(0);
     // Sometimes activation output is bitcast, the actual activation is the
     // second user of the producer of bmm_2's first operand.
```
I think "second user" is a bit confusing, as I think what is meant is "other user". At least the code below shows that we expect that the operand has exactly two users, and we select the other user.

akuegel commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -1205,38 +1326,66 @@ StatusOr<HloInstruction*> FuseFwdMultiHeadedAttentionBlock(
-      activation_output = producer->UserId(activation_output) == 0
-                              ? producer->users()[1]
-                              : producer->users()[0];
+      HloInstruction* bmm2_grad2_user = producer->UserId(activation_output) == 0
```
I think instead of checking producer->UserId you could also check producer->users()[0] == activation_output.
At least I think that is a little bit easier to understand.

akuegel commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -1530,40 +1739,52 @@ StatusOr<bool> CudnnFusedMHARewriter::Run(
+      // We make sure no attention block is matched and replaced twice here
+      if (matched_bmm1.find(matched_result.matched_bmm_1) != matched_bmm1.end())
+        continue;
+      matched_bmm1.insert(matched_result.matched_bmm_1);
```
This can be simplified to:

```
if (!matched_bmm1.insert(matched_result.matched_bmm_1).second) {
        continue;
}
```

Note that our style guide also requires using {} around if blocks if the block is not in the same line as the if. So alternatively you can put the continue also in the same line (if it still fits in 80 characters)

akuegel commented on xla/service/gpu/cudnn_fused_mha_rewriter_test.cc:
```diff
@@ -2893,6 +2893,693 @@ ENTRY main.146 {
+
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str));
```
This looks a bit like an end2end test. If you want an end2end test, I suggest doing that with GetOptimizedModule(), as that will use the GPU pipeline. Otherwise you are not necessarily testing what happens in the real pipeline.
If it should not be an end2end test, I suggest starting off with the HLO after running the passes that you have added here before running CudnnFusedMHARewriter.

Cjkkkk replied:
Not trying to run e2e here. The HLO here are the HLO right before the FMHA pipeline. And we are only running some passes in the FMHA pipeline.

akuegel replied:
So if this should not be an e2e test here, you should remove the passes AlgebraicSimplifier, ReshapeDecomposer, LayoutNormalization, HloCSE down below, and replace the HLO above with what you would get after running those passes.

Cjkkkk replied:
updated.

akuegel commented on xla/service/gpu/cudnn_fused_mha_transpose_fusion.cc:
```diff
@@ -455,9 +458,16 @@ StatusOr<bool> FusePrologueTransposeWithcuDNNFMHA(HloComputation* comp) {
+      if (!config.is_flash_attention()) {
+        TF_ASSIGN_OR_RETURN(changed,
+                            FuseArgPrologueTransposeWithcuDNNFMHA(
+                                fmha, 4, true /*is_lhs*/,
```
Nit: /*is_lhs=*/
(so add the = sign).
Same in the line below.

akuegel commented on xla/service/gpu/tests/gpu_fused_mha_test.cc:
```diff
@@ -2215,6 +2215,442 @@ class MultiHeadedAttentionBMMScaleBiasSoftmaxBMM
+  void TestImpl_Flash_Attention_BMM1_CausalMask_Softmax_BMM2() {
+    stream_executor::CudaComputeCapability cc = GetCudaComputeCapability();
+    se::dnn::VersionInfo real_cudnn_version = GetCudnnVersion();
+    if (!(cc.IsAtLeast(se::CudaComputeCapability::AMPERE) && cc.minor == 0 &&
```
This would potentially result in flaky tests, as it would always pass on GPUs below AMPERE. Ideally we should force the test to always run on AMPERE GPUs. You can see an example here:

https://github.com/openxla/xla/blob/d17c9dd23d76f12013a417923a58e21a27f19b70/xla/service/gpu/BUILD#L3816

Cjkkkk replied:
If I am understanding this correctly, add
```
```

Cjkkkk commented:
@akuegel Hi, Adrian, I am not able to see why `XLA Linux GPU with NVCC` failed. It shows 404 from my side. Could you share some info if you have access to that?

akuegel commented:
> @akuegel Hi, Adrian, I am not able to see why `XLA Linux GPU with NVCC` failed. It shows 404 from my side. Could you share some info if you have access to that?

I also get that error when I click the link. I am not sure where I would be able to see the results. But I will see the results from presubmit testing once this CL is imported, and if any error related to this PR shows up, I will let you know.

akuegel commented:
> Hi, do we have anything pending for this PR?

I would like to wait until the flag change has gone in, so that we get proper testing before this is landed. Otherwise, it might block the flap flip when we suddenly find out it breaks tests.

akuegel commented on xla/service/gpu/tests/gpu_fused_mha_test.cc:
```diff
@@ -2215,6 +2215,442 @@ class MultiHeadedAttentionBMMScaleBiasSoftmaxBMM
+    auto bias_literal = GetInput4DLiteral<T>({2, 6, 2048, 2048}, {3, 2, 1, 0});
+    std::string hlo_string = "";
+    if (std::is_same<T, Eigen::half>::value) {
+      //
```
Is this intentional?

akuegel commented on xla/service/gpu/tests/gpu_fused_mha_test.cc:
```diff
@@ -2215,6 +2215,442 @@ class MultiHeadedAttentionBMMScaleBiasSoftmaxBMM
+    auto do_literal = GetInput4DLiteral<T>({2, 6, 1024, 64}, {3, 2, 1, 0});
+    std::string hlo_string = "";
+    if (std::is_same<T, Eigen::half>::value) {
+      //
```
Is this intentional?

akuegel commented:
> > I would like to wait until the flag change has gone in, so that we get proper testing before this is landed. Otherwise, it might block the flap flip when we suddenly find out it breaks tests.
> 
> This has been merged now.

But I just saw a postsubmit test result that is failing due to the flag flip. We might need to revert again.

akuegel commented on xla/service/gpu/tests/gpu_fused_mha_test.cc:
```diff
@@ -2215,6 +2215,422 @@ class MultiHeadedAttentionBMMScaleBiasSoftmaxBMM
+        GetInput4DLiteral<T>({2, 6, 1024, 64}, {3, 2, 1, 0});
+    std::string hlo_string = "";
+    hlo_string =
+        GetModuleFlash_Attention_Training_BMM1_CausalMask_Softmax_BMM2_HloString_BF16();
```
can you please add a // NOLINT
at the end of the line? It is too long, but I also don't see a way to make it fit into 80 characters without shortening the module description.

akuegel commented on xla/service/gpu/tests/gpu_fused_mha_test.cc:
```diff
@@ -2215,6 +2215,422 @@ class MultiHeadedAttentionBMMScaleBiasSoftmaxBMM
+    auto do_literal = GetInput4DLiteral<T>({2, 6, 1024, 64}, {3, 2, 1, 0});
+    std::string hlo_string = "";
+    hlo_string =
+        GetModuleFlash_Attention_Training_BMM1_Bias_Softmax_BMM2_HloString_BF16();
```
can you please add a // NOLINT
at the end of the line? It is too long, but I also don't see a way to make it fit into 80 characters without shortening the module description.

akuegel commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -1393,67 +1546,72 @@ StatusOr<bool> FuseBwdMultiHeadedAttentionBlock(
+  TF_ASSIGN_OR_RETURN(
+      HloInstruction * bmm_2_grad_1_rhs,
+      ChangeCheckedDimToFastest(comp, bmm_2_grad_1, false /*is_lhs*/,
+                                false /*should_contracting_be_fastest*/));
```
Can you please add

(void)bmm_2_grad_1_rhs;

so that there will be no unused variable warning?

akuegel commented on xla/service/gpu/tests/BUILD:
```diff
@@ -886,10 +886,16 @@ xla_test(
+xla_test(
     name = "gpu_fused_mha_test",
     srcs = ["gpu_fused_mha_test.cc"],
     tags = tf_cuda_tests_tags(),
```
This needs to be removed, the tags will be added automatically for xla_test.

akuegel commented:
FYI: I am still waiting for benchmark results internally, will approve when I have them

Cjkkkk commented:
@akuegel Hey Adrian, any update on the benchmark?

akuegel commented:
> @akuegel Hey Adrian, any update on the benchmark?

Somehow it seems we don't have enough resources to schedule the benchmarking jobs. I still have some pending jobs waiting. This is also already the third time that I try to benchmark it, and the first two attempts resulted in some infrastructure failures. I will give it a try with our other benchmarks, maybe at least those work.

akuegel commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -1381,14 +1528,15 @@ StatusOr<bool> FuseBwdMultiHeadedAttentionBlock(
   HloInstruction* lhs_bmm2_grad_gemm1;
   HloInstruction* rhs_bmm2_grad_gemm2;
   HloInstruction* d_output_grad;
+  HloInstruction* fwd_act;
```
Nit: maybe move this down close to where it is assigned?
I think that makes the code easier to read.

akuegel commented on xla/service/gpu/cudnn_fused_mha_rewriter.cc:
```diff
@@ -1600,21 +1809,33 @@ StatusOr<bool> CudnnFusedMHARewriter::Run(
-      if (!is_mha_module_supported) continue;
 
+      if (!is_mha_module_supported) continue;
+      // flash attention require cuDNN 8.9.3 to run non-fused QKV
```
nit: require -> requires

akuegel reviewed (commented):
I finally got benchmark results, it turns out the issue was fixed recently and I just needed to sync to the latest changes. Benchmarks overall look good.
But I found a few minor things when looking again at the code from this PR, can you please also fix those?

Cjkkkk commented on xla/service/gpu/cudnn_fused_mha_transpose_fusion.cc:
```diff
@@ -137,9 +140,7 @@ StatusOr<bool> FuseArgPrologueTransposeWithcuDNNFMHA(
-      ((transpose_arg_operand->shape().dimensions().at(
-            new_bmm_checked_dims[0]) == 64) ||
-       (IsBwdCustomCallTofMHA(*fmha) && operand_index == 3))) {
+      !(IsBwdCustomCallTofMHA(*fmha) && operand_index == 3)) {
```
for regular non flash attention, the head dim has to be 64 and be the fastest dim. for flash attention, the head dim does not have to be 64(it can be also be 128). Also check fastest dim is 64 or not is not good enough, sequence length can also be 64. The right check would be use the contracting dim of bmm1 as the head dim. I think I remove this constraint as WAR to trigger flash attention fusion but didn't add back the more detailed check here.

Cjkkkk replied:
So after taking the second look, the current checks should work as expected. Since all arguments to fmha calls are already transposed to a layout such that head dim is the fastest dim in the rewriter pass. The checks here only need to make sure head dim is still the fastest before the transpose(this transpose is not the transpose introduced in rewriter, could be other transpose introduced by the framework). `minor_to_major_bmm[0]` is the fastest dim and `new_bmm_checked_dims[0]` is the head dim before transpose. It does not need to be 64, so removing that is necessary.

dathudeptrai commented:
Hi @akuegel , can you share the benchmark results?. I wonder how much the speed-up is generally.

dathudeptrai commented:
@akuegel Thanks. BTW, is that true that in the near future, in TF side, using `jit_complie=True` in tf.function will try to apply flash attention?.

akuegel commented:
> @akuegel Thanks. BTW, is that true that in the near future, in TF side, using `jit_complie=True` in tf.function will try to apply flash attention?.

The flag is already enabled by default, so at least for HLO that matches what is currently supported it should already work.

akuegel commented:
@Cjkkkk We get a test failure for FlashAttentionBMMScaleCausalMaskSoftmaxBMM.Flash_Attention_BMM1_CausalMask_Softmax_BMM2_BF16

Can you please check whether you can reproduce this?

Cjkkkk commented:
> FlashAttentionBMMScaleCausalMaskSoftmaxBMM.Flash_Attention_BMM1_CausalMask_Softmax_BMM2_BF16

Sorry for the late reply, was OOO since last Wednesday. I can't reproduce this on A100 + cuDNN 8.9.7. Could you share the error message and the GPU/cuDNN version?

Cjkkkk commented:
@akuegel Hey Adrian, just a kindly reminder that let me know if you have the error message and the GPU/cuDNN version of the test failure. Thanks!

akuegel commented:
> @akuegel Hey Adrian, just a kindly reminder that let me know if you have the error message and the GPU/cuDNN version of the test failure. Thanks!

I was out on vacation, and will have 2 more days of vacation on Wednesday and Thursday.
GPU is A100, cuDNN version is 8.9.4
We get a SIGABRT in ```stream_executor::gpu::(anonymous namespace)::(anonymous namespace)::GetCudnnFlashAttentionOperationGraph()```

Cjkkkk commented:
> Flash_Attention_BMM1_CausalMask_Softmax_BMM2_BF16

I still can't reproduce this with A100 + cuDNN 8.9.4. Could we get more info like call stack to see where SIGABRT happened.

akuegel commented:
> > Flash_Attention_BMM1_CausalMask_Softmax_BMM2_BF16
> 
> I still can't reproduce this with A100 + cuDNN 8.9.4. Could we get more info like call stack to see where SIGABRT happened.

Unfortunately the internal version of this PR has been garbage collected by now. I will approve this PR again so that we get a new internal version.

dathudeptrai commented:
@Cjkkkk @akuegel will this flash-attention support attention-bias format like in T5 ?

akuegel commented:
> > I still can't reproduce this with A100 + cuDNN 8.9.4. Could we get more info like call stack to see where SIGABRT happened.
> 
> Unfortunately the internal version of this PR has been garbage collected by now. I will approve this PR again so that we get a new internal version.

Here is the full stack trace, but I am not sure whether it helps. When you try to reproduce, are you running with tensorflow head, or do you have some custom patches and/or non-default flags?

```
[ RUN      ] FlashAttentionBMMScaleCausalMaskSoftmaxBMM.Flash_Attention_BMM1_CausalMask_Softmax_BMM2_BF16
*** SIGABRT received by PID 5331 (TID 5331) on cpu 58 from PID 5331; stack trace: ***
PC: @     0x7fb4b0430347  (unknown)  gsignal
    @     0x7fb0e50a579a        944  FailureSignalHandler()
    @     0x7fb4b88051c0  (unknown)  (unknown)
    @     0x7fb4b0430347        136  gsignal
    @     0x7fb4b0431797        304  abort
    @     0x7fb4432da80d      13856  stream_executor::gpu::(anonymous namespace)::(anonymous namespace)::GetCudnnFlashAttentionOperationGraph()
    @     0x7fb4432d5061       2160  stream_executor::gpu::CudnnSupport::FusedMHARunnerFromDesc()
    @     0x7fb4682e4da5        416  stream_executor::Stream::FusedMHARunnerFromDesc()
    @     0x7fb4682e48c3        416  stream_executor::dnn::FusedMHAOp::RunnerFromAlgorithmDesc()
    @     0x7fb4682e4419        336  stream_executor::dnn::LazyOpRunner<>::GetOrCreateRunner()::{lambda()#1}::operator()()
    @     0x7fb4682e418d         48  absl::base_internal::CallOnceImpl<>()
    @     0x7fb4682dd496       1504  xla::gpu::RunGpuFMHA()
    @     0x7fb46873f93f       6768  xla::gpu::FusedAttentionForwardImpl()
    @     0x7fb46873a919       1968  xla::runtime::CustomCallHandler<>::call()
    @     0x7fb46872e277       1040  xla::gpu::FusedAttentionSoftmaxInference()
    @     0x7fb4af1916e3        848  (unknown)
    @     0x7fb4509ed999        128  xla::runtime::Executable::Execute()
    @     0x7fb46c96f2e8       1584  xla::gpu::GpuRuntimeExecutable::Execute()
    @     0x7fb46d1c8215        608  xla::gpu::GpuExecutable::ExecuteThunksOrXlaRuntime()
    @     0x7fb46d1c67b0       2048  xla::gpu::GpuExecutable::ExecuteAsyncOnStreamImpl()
    @     0x7fb46d1c616d         80  xla::gpu::GpuExecutable::ExecuteAsyncOnStream()
    @     0x7fb41bc4b277        976  xla::Executable::ExecuteAsyncOnStreamWrapper()
    @     0x7fb41bc4afc5        944  xla::Executable::ExecuteOnStreamWrapper()
    @     0x7fb5c59de655       2208  xla::HloRunner::ExecuteWithExecutionInputs()
    @     0x7fb5c59df60c       2336  xla::HloRunner::ExecuteWithMovedDeviceBuffers()
    @     0x7fb5c59dccd8        112  xla::HloRunner::ExecuteWithMovedDeviceBuffersAndBufferAssignment()
    @     0x7fb5c59dc91c       1888  xla::HloRunner::Execute()
    @     0x7fb5c723704d        112  xla::HloTestBase::ExecuteAndTransfer()
    @     0x7fb5caba3402       3648  xla::gpu::MultiHeadedAttentionTest::ExecuteAndCompare()
    @     0x7fb5cab9c562       1664  xla::gpu::FlashAttentionBMMScaleCausalMaskSoftmaxBMM::TestImpl_Flash_Attention_BMM1_CausalMask_Softmax_BMM2<>()
    @     0x7fb5bbef1bf7         48  testing::Test::Run()
    @     0x7fb5bbef2e89         64  testing::TestInfo::Run()
    @     0x7fb5bbef3f53         96  testing::TestSuite::Run()
    @     0x7fb5bbf0618c        192  testing::internal::UnitTestImpl::RunAllTests()
    @     0x7fb5bbf05154         64  testing::UnitTest::Run()
    @     0x7fb5bf3f984d        176  main
    @     0x7fb4b041c633        192  __libc_start_main
    @     0x55dd2bdc792a  (unknown)  [../sysdeps/x86_64/start.S:120](https://cs.corp.google.com/piper///depot/sysdeps/x86_64/start.S?l=120&ws=tap-presubmit-server/143975588&snapshot=2) _start
```

Cjkkkk commented:
> > > I still can't reproduce this with A100 + cuDNN 8.9.4. Could we get more info like call stack to see where SIGABRT happened.
> > 
> > 
> > Unfortunately the internal version of this PR has been garbage collected by now. I will approve this PR again so that we get a new internal version.
> 
> Here is the full stack trace, but I am not sure whether it helps. When you try to reproduce, are you running with tensorflow head, or do you have some custom patches and/or non-default flags?
> 
> ```
> [ RUN      ] FlashAttentionBMMScaleCausalMaskSoftmaxBMM.Flash_Attention_BMM1_CausalMask_Softmax_BMM2_BF16
> *** SIGABRT received by PID 5331 (TID 5331) on cpu 58 from PID 5331; stack trace: ***
> PC: @     0x7fb4b0430347  (unknown)  gsignal
>     @     0x7fb0e50a579a        944  FailureSignalHandler()
>     @     0x7fb4b88051c0  (unknown)  (unknown)
>     @     0x7fb4b0430347        136  gsignal
>     @     0x7fb4b0431797        304  abort
>     @     0x7fb4432da80d      13856  stream_executor::gpu::(anonymous namespace)::(anonymous namespace)::GetCudnnFlashAttentionOperationGraph()
>     @     0x7fb4432d5061       2160  stream_executor::gpu::CudnnSupport::FusedMHARunnerFromDesc()
>     @     0x7fb4682e4da5        416  stream_executor::Stream::FusedMHARunnerFromDesc()
>     @     0x7fb4682e48c3        416  stream_executor::dnn::FusedMHAOp::RunnerFromAlgorithmDesc()
>     @     0x7fb4682e4419        336  stream_executor::dnn::LazyOpRunner<>::GetOrCreateRunner()::{lambda()#1}::operator()()
>     @     0x7fb4682e418d         48  absl::base_internal::CallOnceImpl<>()
>     @     0x7fb4682dd496       1504  xla::gpu::RunGpuFMHA()
>     @     0x7fb46873f93f       6768  xla::gpu::FusedAttentionForwardImpl()
>     @     0x7fb46873a919       1968  xla::runtime::CustomCallHandler<>::call()
>     @     0x7fb46872e277       1040  xla::gpu::FusedAttentionSoftmaxInference()
>     @     0x7fb4af1916e3        848  (unknown)
>     @     0x7fb4509ed999        128  xla::runtime::Executable::Execute()
>     @     0x7fb46c96f2e8       1584  xla::gpu::GpuRuntimeExecutable::Execute()
>     @     0x7fb46d1c8215        608  xla::gpu::GpuExecutable::ExecuteThunksOrXlaRuntime()
>     @     0x7fb46d1c67b0       2048  xla::gpu::GpuExecutable::ExecuteAsyncOnStreamImpl()
>     @     0x7fb46d1c616d         80  xla::gpu::GpuExecutable::ExecuteAsyncOnStream()
>     @     0x7fb41bc4b277        976  xla::Executable::ExecuteAsyncOnStreamWrapper()
>     @     0x7fb41bc4afc5        944  xla::Executable::ExecuteOnStreamWrapper()
>     @     0x7fb5c59de655       2208  xla::HloRunner::ExecuteWithExecutionInputs()
>     @     0x7fb5c59df60c       2336  xla::HloRunner::ExecuteWithMovedDeviceBuffers()
>     @     0x7fb5c59dccd8        112  xla::HloRunner::ExecuteWithMovedDeviceBuffersAndBufferAssignment()
>     @     0x7fb5c59dc91c       1888  xla::HloRunner::Execute()
>     @     0x7fb5c723704d        112  xla::HloTestBase::ExecuteAndTransfer()
>     @     0x7fb5caba3402       3648  xla::gpu::MultiHeadedAttentionTest::ExecuteAndCompare()
>     @     0x7fb5cab9c562       1664  xla::gpu::FlashAttentionBMMScaleCausalMaskSoftmaxBMM::TestImpl_Flash_Attention_BMM1_CausalMask_Softmax_BMM2<>()
>     @     0x7fb5bbef1bf7         48  testing::Test::Run()
>     @     0x7fb5bbef2e89         64  testing::TestInfo::Run()
>     @     0x7fb5bbef3f53         96  testing::TestSuite::Run()
>     @     0x7fb5bbf0618c        192  testing::internal::UnitTestImpl::RunAllTests()
>     @     0x7fb5bbf05154         64  testing::UnitTest::Run()
>     @     0x7fb5bf3f984d        176  main
>     @     0x7fb4b041c633        192  __libc_start_main
> ```

I rebased on the XLA head last Friday and rerun again and all tests passed. This is the head commit I rebased on.
```
commit a8531b6ad729a4cdff5aa7a37f56fd529a5bbd86 (origin/main, origin/HEAD, main)
Author: A. Unique TensorFlower <gardener@tensorflow.org>
Date:   Fri Jan 5 08:55:53 2024 -0800

    [Service] Export Service::BuildExecutable for usage by library users
    
    PiperOrigin-RevId: 596007039
```

This is the command I used: 
```
TF_CUDA_COMPUTE_CAPABILITIES="sm_70,sm_75,compute_80" bazel test --config=cuda --local_ram_resources=HOST_RAM*.5  --local_cpu_resources=24 --nocheck_visibil
ity --copt="-Wno-error=switch" --java_runtime_version=remotejdk_11 --test_output=all //xla/service/gpu/tests:gpu_fused_mha_test
```
I didn't use any other XLA_FLAGS besides the ones set in the unit test. `debug_options.set_xla_gpu_enable_cudnn_fmha(true);`

Is this the latest result after I rebased last Friday?

Cjkkkk commented:
> @Cjkkkk @akuegel will this flash-attention support attention-bias format like in T5 ?

Yes, t5x is supported last time I checked. For now, the sequence <= 512 will not trigger cuDNN flash attention but cuDNN fused attention.

akuegel commented:
> This is the command I used:
> 
> ```
> TF_CUDA_COMPUTE_CAPABILITIES="sm_70,sm_75,compute_80" bazel test --config=cuda --local_ram_resources=HOST_RAM*.5  --local_cpu_resources=24 --nocheck_visibil
> ity --copt="-Wno-error=switch" --java_runtime_version=remotejdk_11 --test_output=all //xla/service/gpu/tests:gpu_fused_mha_test
> ```
> 
> I didn't use any other XLA_FLAGS besides the ones set in the unit test. `debug_options.set_xla_gpu_enable_cudnn_fmha(true);`
> 
> Is this the latest result after I rebased last Friday?

Yes, this is the latest result based on the import when I approved the PR again. I see that the open source version of the test worked fine (XLA Linux GPU CI passed, and I see this specific test passing in that run). So apparently in our internal setup, something must be different. In that case, we need to figure this out internally.

Cjkkkk commented:
@akuegel hey Adrian, seems like the issue is with the optional deference not checking null opt, I observed there is a dropout_rate might be nullopt and I added guards for that. Could you rerun the internal tests to see if the issues are gone?

akuegel commented:
> @akuegel hey Adrian, seems like the issue is with the optional deference not checking null opt, I observed there is a dropout_rate might be nullopt and I added guards for that. Could you rerun the internal tests to see if the issues are gone?

Thanks for the fix, it seems quite likely that this was the issue. Can you please rebase? It seems now there is a merge conflict.
